File: go/day3/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"regexp"
	"strconv"
	"strings"
)

func extractNumbers(expression string) (int, int) {
	re := regexp.MustCompile(`\d+,\d+`)
	match := re.FindString(expression)
	if match == "" {
		return 0, 0
	}
	parts := strings.Split(match, ",")
	first, _ := strconv.Atoi(parts[0])
	second, _ := strconv.Atoi(parts[1])
	return first, second
}

func evaluateLine(line string) int {
	re := regexp.MustCompile(`mul\(\d+,\d+\)`) // Matches expressions like mul(2,3)
	matches := re.FindAllString(line, -1)
	sum := 0

	for _, expr := range matches {
		if len(expr) <= 12 {
			num1, num2 := extractNumbers(expr)
			sum += num1 * num2
		}
	}

	return sum
}

func partOne() int {
	file, err := os.Open("../../input/day3.txt")
	if err != nil {
		panic(err)
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	total := 0

	for scanner.Scan() {
		total += evaluateLine(scanner.Text())
	}

	if err := scanner.Err(); err != nil {
		panic(err)
	}

	return total
}

func partTwo() int {
	file, err := os.Open("../../input/day3.txt")
	if err != nil {
		panic(err)
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)

	var lines []string
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}

	if err := scanner.Err(); err != nil {
		panic(err)
	}

	dos := []string{}
	for _, line := range lines {
		parts := strings.Split(line, "do()")
		for _, part := range parts {
			doPart := strings.Split(part, "don't()")[0]
			dos = append(dos, doPart)
		}
	}

	result := 0
	for _, line := range dos {
		result += evaluateLine(line)
	}

	return result
}

func main() {
	fmt.Println(partTwo())
}

File: python/day3/main.py
import re
import numpy as np


inp = open("../../input/day3.txt", "r")

def partOne():
    res = 0
    for line in inp:
        for y in re.findall(r"mul+\(+[0-9]+,[0-9]+\)", line):
            if y[0]+y[1]+y[2] == "mul" and len(y[3:]) <= 12:
                val = re.findall(r"[0-9]+,[0-9]+",re.findall(r"\([^)]*\)",y[3:])[0])
                if len(val) > 0:
                    cont = val[0] + ""
                    values = cont.split(",")
                    reslt = int(values[0]) * int(values[1])
                    res = res + reslt
    return res
# 173785482

def partTwo():
    res = 0
    lines = []
    for line in inp:
        l = line.split("do()")
        lines.append(l)
    dos = []
    for line in lines:
        for l in line:
            dos.append(l.split("don't()")[0])
    for line in dos:
        for y in re.findall(r"mul+\(+[0-9]+,[0-9]+\)", line):
            if y[0]+y[1]+y[2] == "mul" and len(y[3:]) <= 12:
                val = re.findall(r"[0-9]+,[0-9]+",re.findall(r"\([^)]*\)",y[3:])[0])
                if len(val) > 0:
                    cont = val[0] + ""
                    values = cont.split(",")
                    reslt = int(values[0]) * int(values[1])
                    res = res + reslt
    return res
print(partTwo())

'''

982038u40812u01023
921380948019389

0213941234don128374089178230948019234
032837273279

89213749023484983

902380923590
do2384950824

39857295dont90384094238059


'''